tests/test_dataclasses.py

```diff
@@ -136,9 +136,18 @@ def double_a(cls, v: int) -> int:
 @pytest.mark.parametrize(
     'config',
     [
+        ConfigDict(validate_assignment=True),
         ConfigDict(validate_assignment=False),
+        ConfigDict(extra=None),
+        ConfigDict(extra=Extra.forbid),
         ConfigDict(extra=Extra.ignore),
+        ConfigDict(validate_assignment=False, extra=None),
+        ConfigDict(validate_assignment=False, extra=Extra.forbid),
```

Comment:
This is a weird scenario: even though `extra=Extra.forbid` is set we don't actually forbid the extra because `validate_assignment` is False. This applies to BaseModel as well I think.

I'm not sure what the right thing to do is here, maybe `validate_assignment` needs to be some sort of enum? Or maybe we need separate `extra` settings for construction/validation and assignment?